### PR TITLE
Experimental cleanup of abs

### PR DESCRIPTION
**Pull request by mkmik**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -98,19 +98,22 @@ impl Index {
...
-                }
-            }
+            &Head(n) => ensure_within(0, n, len),
+            &Tail(n) => ensure_within(0, n, len).map(|n| len - n),
             Default => None,
         }
     }
 }
 
+/// Return Some(n) if min <= None < max, otherwise return None.
```
Comment (glyn) [inline src/lib.rs]:
```suggestion
/// Return Some(n) if min <= n <= max, otherwise return None.
```

File: src/lib.rs [hunk 2]
```diff
@@ -98,19 +98,22 @@ impl Index {
...
+fn ensure_within(min: usize, n: usize, max: usize) -> Option<usize> {
+    if n < min || max < n {
+        None
+    } else {
+        Some(n)
+    }
```
Comment (glyn) [inline src/lib.rs]: Why not make this match the doc comment?
```suggestion
    if min <= n && n <= max {
        Some(n)
    } else {
        None
    }
```

**Review by glyn — CHANGES_REQUESTED** [review]
I like the direction, but suggested some changes.

File: src/lib.rs [hunk 3]
```diff
@@ -90,27 +90,30 @@ impl Slice {
...
-                }
-            }
+            &Head(n) => ensure_within(0, n, len),
+            &Tail(n) => ensure_within(0, n, len).map(|n| len - n),
             Default => None,
         }
     }
 }
 
+/// Return Some(n) if min <= n < max, otherwise return None.
```
Comment (mkmik) [inline src/lib.rs]: I think having non-inclusive end range is more clear (and we can also use std::ops::Range)

Comment (mkmik) [reply src/lib.rs]: @glyn PTAL

**Review by glyn — APPROVED** [review]
LGTM, thanks.

**Pull request merged**